datafusion/expr/src/udaf.rs

```diff
@@ -507,6 +510,21 @@ pub trait AggregateUDFImpl: Debug + Send + Sync {
     fn coerce_types(&self, _arg_types: &[DataType]) -> Result<Vec<DataType>> {
         not_impl_err!("Function {} does not implement coerce_types", self.name())
     }
+
+    /// Dynamic equality. Allows customizing the equality of aggregate UDFs.
+    /// By default, compares the UDF name and signature.
+    fn equals(&self, other: &dyn AggregateUDFImpl) -> bool {
+        self.name() == other.name() && self.signature() == other.signature()
+    }
+
+    /// Dynamic hashing. Allows customizing the hash code of aggregate UDFs.
+    /// By default, hashes the UDF name and signature.
+    fn hash_value(&self) -> u64 {
```

Comment:
I think it would be good to note here that eq and hash value need to be consistent.

Something like this perhaps:

```suggestion
    /// Returns a hash value for this aggregate UDF. 
    ///     
    /// Allows customizing the hash code of aggregate UDFs. Similarly to 
    /// [`std::hash::Hash`], if [`Self::equals`] 
    /// returns true for two aggregate UDFs, the value of `hash_value` must as well. 
    /// 
    /// By default, hashes [`Self::name`] and [`Self::signature`]
    fn hash_value(&self) -> u64 {
```